src/core/vmm.d

```diff
@@ -0,0 +1,485 @@
+}
+
+/**
+    Resets a set of pages that form the specified $(D range) of virtual memory.
+    This instructs an OS that data in the pages is no loger of interest to the application,
+    so it's free to reuse any of page involved while keeping mapping intact.
+    Subsequent access will result in either reading a prior data or a fresh page.
+
+*/
+static bool reset(void[] range)
+{
+    return osReset(range.ptr, range.length);
+}
+
+
+/**
+    Decommit a set of pages that form the specified $(D range) of virtual memory.
+    While system retains the address space range, accessing memory inside of it
+    is illegal. The exact consequences of such illegal access are platform
+    dependent.
+
+*/
+static bool decommit(void[] range)
+{
+    return osDecommit(range.ptr, range.length);
+}
+
+/**
+    Decommit and releases pages that were previously allocated and/or reserved.
+    $(D range) must contain exactly the same address range as the one
+    returned by $(LREF allocate) or $(LREF reserve).
+
+    It doesn't matter what the state of each page was (reserved vs commited),
+    the OS decommits pages as needed.
+*/
+static bool free(void[] range)
+{
+    return osUnmap(range.ptr, range.length);
+}
+
+/**
+    Set permissions on the set of pages that form $(D range).
+*/
+static bool protect(void[] range, MemoryAccess access)
+{
+    return osProtect(range.ptr, range.length, access);
+}
+
+/**
+    Lock pages that form $(D range) to prevent their paging out.
+*/
+static bool lock(void[] range)
+{
+    return osLockPages(range.ptr, range.length);
+}
+
+/**
+    Undo the effect of locking (with $(LREF lock)) pages that form $(D range).
+*/
+static bool unlock(void[] range)
+{
+    return osUnlockPages(range.ptr, range.length);
+}
+
+/**
+    Round up a pointer or integral $(D arg)
+    to the nearest multiple of platform's page size.
+*/
+T roundUpToPage(T)(T arg)
+    if (is(T : void*) || is(T : size_t))
+{
+    return cast(T)((cast(size_t)arg + pageSize-1) & ~(pageSize-1));
+}
+
+/**
+    Round down a pointer or integral $(D arg)
+    to the nearest multiple of platform's page size.
+*/
+T roundDownToPage(T)(T arg)
+    if (is(T : void*) || is(T : size_t))
+{
+    return cast(T)(cast(size_t)arg & ~(pageSize-1));
+}
+
+//code to initialize pageSize constant
+version (Windows)
+{
+    import core.sys.windows.windows;
+
+    alias int pthread_t;
+
+    pthread_t pthread_self()
+    {
+        return cast(pthread_t) GetCurrentThreadId();
+    }
+
+
+    shared static this()
+    {
+        SYSTEM_INFO sysinfo;
+        GetSystemInfo(&sysinfo);
+        pageSize = sysinfo.dwPageSize;
+    }
+}
+else version (Posix)
+{
+    import core.sys.posix.unistd;
+
+    private shared int fzero;
+
+    shared static this()
+    {
+        pageSize = sysconf(_SC_PAGESIZE);
+        version(linux) //temporary
+        {
+            import core.sys.posix.fcntl;
+            fzero = open("/dev/zero", O_RDWR);
+            assert(fzero != -1);
```

Comment:
Doesn't mmap come with a private, not file backed option?
How about MAP_ANONYMOUS or fd == -1?